### PR TITLE
Add decimalyear format to Time

### PR DESCRIPTION
**Pull request by taldcroft**
This is a replacement for #2788 which uses a more efficient method to do the transformation.

### CODE REVIEW

**Comment by taldcroft** [comment]
cc @mhvk 
Closes #2788.

File: astropy/time/core.py [hunk 1]
```diff
@@ -1412,6 +1413,68 @@ def value(self):
         return (self.jd1 - erfa.DJM0) + self.jd2
 
 
+class TimeDecimalYear(TimeFormat):
+    """
+    Time as a decimal year, with integer values corresponding to midnight
+    of the first day of each year.  For example 2005.5 is corresponds to the
```
Comment (mhvk) [inline astropy/time/core.py]: This example is for 2005!  Should be `2000.5`

File: astropy/time/core.py [hunk 2]
```diff
@@ -1412,6 +1413,68 @@ def value(self):
...
+    ISO time '2000-07-02 00:00:00'.
+    """
+    name = 'decimalyear'
+
+    def set_jds(self, val1, val2):
+        self._check_scale(self._scale)  # Validate scale.
+
+        val = (val1 + val2).astype(np.double)
+        print(val)
```
Comment (mhvk) [inline astropy/time/core.py]: Remove print statement

Comment (taldcroft) [inline astropy/time/core.py]: Maybe I shouldn't be coding before 8am.  :smile:

File: astropy/time/core.py [hunk 3]
```diff
@@ -1412,6 +1413,68 @@ def value(self):
...
+        iy_start = np.trunc(val).astype(np.int)
```
Comment (mhvk) [inline astropy/time/core.py]: We can do the split in years and fraction to somewhat better precision with a short version of `day_frac`:
```
sum12, err12 = two_sum(val1, val2)
iy_start = np.trunc(sum12).astype(np.int)
extra, y_frac = two_sum(sum12, -iy_start)
y_frac += extra + err12 
```

File: astropy/time/core.py [hunk 4]
```diff
@@ -1412,6 +1413,68 @@ def value(self):
...
+
+        imon = np.ones_like(iy_start)
+        iday = np.ones_like(iy_start)
+        ihr = np.zeros_like(iy_start)
+        imin = np.zeros_like(iy_start)
+        isec = np.zeros_like(val)
```
Comment (mhvk) [inline astropy/time/core.py]: If we go with the above, this would be `np.zeros_like(y_frac)`

File: astropy/time/core.py [hunk 5]
```diff
@@ -1412,6 +1413,68 @@ def value(self):
...
+        isec = np.zeros_like(self.jd1)
+
+        # Possible enhancement: use np.unique to only compute start, stop
+        # for unique values of iy_start.
+        jd1_start, jd2_start = erfa_time.dtf_jd(
+            self.scale.upper().encode('utf8'), iy_start, imon, iday, ihr, imin, isec)
+        jd1_end, jd2_end = erfa_time.dtf_jd(
+            self.scale.upper().encode('utf8'), iy_start + 1, imon, iday, ihr, imin, isec)
+
+        jd_start = jd1_start + jd2_start
```
Comment (mhvk) [inline astropy/time/core.py]: It's no big deal but perhaps again try to keep precision as best as possible with:
```
dt = (self.jd1 - jd1_start) + (self.jd2 - jd2_start)
dt_end = (jd1_end - jd1_start) + (jd2_end - jd2_start)
decimalyear = iy_start + dt / dt_end
```

**Comment by mhvk** [comment]
@taldcroft - looks good! My precision-keeping comments are possibly a bit superfluous, but one might as well (especially the one for the output). With those, I'm happy for this to be merged.

**Comment by taldcroft** [comment]
@mhvk - thanks for the careful review and good suggestions on precision!

File: astropy/time/core.py [hunk 6]
```diff
@@ -1412,6 +1413,69 @@ def value(self):
...
+        ihr = np.zeros_like(iy_start)
+        imin = np.zeros_like(iy_start)
+        isec = np.zeros_like(self.jd1)
+
+        # Possible enhancement: use np.unique to only compute start, stop
+        # for unique values of iy_start.
+        jd1_start, jd2_start = erfa_time.dtf_jd(
+            self.scale.upper().encode('utf8'), iy_start, imon, iday, ihr, imin, isec)
+        jd1_end, jd2_end = erfa_time.dtf_jd(
+            self.scale.upper().encode('utf8'), iy_start + 1, imon, iday, ihr, imin, isec)
```
Comment (embray) [inline astropy/time/core.py]: ISTM there's some repetition between here, and the same code in `set_jds` that could be factored out.

**Pull request merged**